twilio-labs/open-pixel-art
# Add merge driver

## Checklist



- [x] I ran `npm test` locally and it passed without errors.

- [ ] I only edited the `_data/pixels.json` file.

- [ ] I entered the `username` in the `pixels.json` that I'm also using to create this pull request.
- [x] I acknowledge that all my contributions will be made under the project's [license](../../LICENSE.md).










## Description

This PR is adding a custom merge driver for the `pixels.json` as well as a sync script that would automatically sync up your fork with the changes from the main repository.

Some related design decisions:
- If the pixel someone tried to contribute has already been contributed, the next possible pixel will be picked by increasing the x coordinate and then afterwards the y coordinate if necessary. The user will see a message and can decide to change it afterwards.
- If there are more than 1 pixel change in the merge conflict the changes from `current` will all be taken and the conflicting pixels will be written into a `pixels-conflict.log` file at the root of the project with the following content:

```
The following pixels were removed from the _data/pixels.json file due to a conflict. 
If this was by accident please place them back into the right location and commit again.

{
  "data": [
    {"y": 10, "x": 17, "color": "#000000", "username": "pandabear"},
    {"y": 12, "x": 23, "color": "#000000", "username": "dom-twilio"}
  ]
}
```

## Testing

### Installing the merge driver:

After checking out this branch, run `npm install` and you should see in `.git/config` the following:

```
[merge "pixels"]
	name = A custom merge driver for the pixels file
	driver = node scripts/mergePixels.js %O %A %B
```

### Testing the merge driver:

From this branch run:
```bash
git checkout -b test1
# add a pixel...
git add _data/pixels.json
git commit -m "update pixels.json"
# git checkout [this branch]
git checkout -b test2
# add the same pixel coordinates but with a different username
git add _data/pixels.json
git commit -m "update pixels.json"
git merge test1
# check out _data/pixels.json and your pixel from test2 should have moved to the next free spot
```

### Testing the sync script

In a cloned repository with a fork as `origin` checkout a new branch from `master` and do some changes. Commit them and run `npm run sync-fork`. This should:
1. Add a remote called `twilio-labs`
2. Fetch the latest changes from `twilio-labs/master`
3. Merge them into your `master`
4. Push them to your fork on GitHub.com
5. Check your branch again 
6. Merge the changes from `master`

### Running tests

Run `npm test`

## Related issues

#580
